feat: allow to edit and entry

Jeansidharta opened the pull request:
Closes #91.

This allows the user to edit an entry in telescope. To edit an entry, you have to open telescope, use the edit mapping (default <C-e> in insert mode), edit the new buffer however you want, then close the buffer (`:q`) to save the changes. Here are some screenshots:

- Open telescope: ![image](https://user-images.githubusercontent.com/26410371/212742692-9a52ca83-4f4f-43ce-aec0-a3379373f2fa.png)
- Press <C-e> in insert mode: ![image](https://user-images.githubusercontent.com/26410371/212742733-a0db01ce-66de-4330-9e61-8d1ac17bf648.png)
- Change the buffer contents: ![image](https://user-images.githubusercontent.com/26410371/212742796-625fcc9e-c541-4ccb-b890-71d2ed293696.png)
- Exit the new buffer (`:q`):  ![image](https://user-images.githubusercontent.com/26410371/212742855-6eb662d7-16de-4819-82ee-d14cc8a6ffd8.png)

AckslD commented on lua/neoclip/sorted_set.lua:
```diff
@@ -47,6 +47,16 @@ local insert = function(self, entry, opts)
     end
 end
 
+local edit = function(self, entry, newEntry)
```
Would it make more sense to call this `replace`?

Jeansidharta replied:
Yes. I like that name more

AckslD commented on lua/neoclip/sorted_set.lua:
```diff
@@ -47,6 +47,16 @@ local insert = function(self, entry, opts)
+    node.value = newEntry -- Replace entry in node.
+
+    self.entries[key] = nil
+    self.entries[new_key] = node
```
Don't we also need to update the underlying linked list? (`self.ll`)

Jeansidharta replied:
The idea behind this piece of code is to only replace the contents of the node, and not the node itself. This way, we won't change the order of the nodes.

The reason this function is that complex is because I also have to update the dictionary's reference to the nodes, using the new calculated hash for the node's content.

AckslD replied:
Hmm, but the linked list node also holds a value (which is used by pop for example). Could maybe change but at the moment I think also the value of the linked list node need to be updated. Could you maybe also add some test case for that?

Jeansidharta replied:
I think I'm a bit confused.

The `self.entries` dictionary has the linked list's nodes as values (as seen [here](https://github.com/AckslD/nvim-neoclip.lua/blob/main/lua/neoclip/sorted_set.lua#L27)). So doing `self.entries[key].value` is already updating the value of the linked list's node, no?

And yeah, I like the idea of adding some test cases. I'll give it a try

Jeansidharta replied:
Hey @AckslD I just added some tests for this new `replace` function. Let me know what you think

Jeansidharta commented on lua/neoclip/sorted_set.lua:
```diff
@@ -47,6 +47,21 @@ local insert = function(self, entry, opts)
+    node.value = newEntry -- Replace entry in node.
+
+    -- Prevent duplicate entries
+    if self.entries[new_key] and key ~= new_key then
+        self:remove(newEntry)
+    end
```
While implementing the new tests, I realized my implementation allowed for duplicate entries to exist in the linked list. This is a problem because they will both have the same hash, and the dictionary won't be able to track them both. These three lines fix that

AckslD replied:
Cool!

AckslD commented on lua/neoclip/sorted_set.lua:
```diff
@@ -47,6 +47,21 @@ local insert = function(self, entry, opts)
     end
 end
 
+local replace = function(self, entry, newEntry)
```
```suggestion
local replace = function(self, entry, new_entry)
```
just to keep the naming convention as lower_case

Jeansidharta replied:
Nice catch!

AckslD commented on lua/neoclip/sorted_set.lua:
```diff
@@ -47,6 +47,21 @@ local insert = function(self, entry, opts)
+local replace = function(self, entry, newEntry)
+    local key = hash(entry)
+    local new_key = hash(newEntry)
+    local node = self.entries[key];
```
should the case when the `entry` is not already in the set be handled somehow?

Jeansidharta replied:
I think it's a good idea, but no other function has error handling. so I'm not sure how I'd got about that. I think I'll just call `error` with a good error message. If you have a better suggestion, let me know!

AckslD replied:
I so far used [this warn function] for errors. Do you know if either has advantage? I haven't used `error` before?

Jeansidharta replied:
I see. the `error` function is a little different, in that it will guarantee a stop in the current stack execution. It acts like a `throw` in another languages (the equivalent `catch` would be a `pcall`). This can be both good and bad. [Here's](https://github.com/AckslD/nvim-neoclip.lua/blob/main/lua/neoclip/db.lua#L8) an example i which stopping the current lua execution would be really bad. But I think sometimes it can be good to make sure the user can't do anything else if the current function fails.

In this case, I think it makes sense to throw a proper error, indicating that the function failed to do it's job. This may prevent the user from assuming the entry has been edited when it has not. What are your thoughts on that?

AckslD commented on README.md:
```diff
@@ -287,6 +287,9 @@ If you don't want to use the setting `continuous_sync`, but still keep two insta
 ### Remove entries
 You can remove entries manually using the keybinds for `delete`. You can also delete the whole history with `:lua require('neoclip').clear_history()`.
 
+### Replace entries
```
In terms of usability this seems quite low-level. Maybe it would be nice if there could be a keymap in the picker which opens the entry in a buffer which you can edit and when you save it it gets replaced. Alternatively maybe just some function that asks the user what entry to replace and what the new one should be using `vim.ui.input`? The former way seems nicer though as you can more easily actually "edit" an entry rather than replacing it.

Not sure if this goes beyond what you wanted to do but seems like just having the `replace` function might not be so usable?

Jeansidharta replied:
The former way is exactly the thing I was trying to implement when I first created this issue (#91). I didn't add it to this commit in fear of being too specific to my needs. I'll move my code over here. and you can take a look at how it's being done

Jeansidharta replied:
@AckslD It's done! I've implemented the edit function in [this](https://github.com/AckslD/nvim-neoclip.lua/pull/93/commits/ccd7e713163a60c818c2e4159a92d78b072007d0) commit.

I've mapped it to <c-e> in telescope, and called it "edit" (which to me just seems more intuitive than replace). It opens a new floating window with the entry's content and filetype set. Whenever the window is closed, the old entry's content is replaced, and telescope is reopened with the new entry pre-selected.

Take a look and let me know what you think!

Jeansidharta commented:
Hey, I noticed [this](https://github.com/AckslD/nvim-neoclip.lua/blob/main/lua/neoclip/telescope.lua#L57-L58) and [this](https://github.com/AckslD/nvim-neoclip.lua/blob/main/lua/neoclip/telescope.lua#L41-L42) todos could possibly be solved with [this](https://github.com/AckslD/nvim-neoclip.lua/pull/93/files#diff-1bdfb6249c123cbeb779faa6aa0f693fe2e2c19993cf1af4e4da224dd39ead8aR17-R29) new function that I just wrote. Instead of trying to actions these functions on the buffer "behind" telescope, we could close telescope, apply the actions, and reopen telescope. What do you think? I'd be willing to open another PR, after this one is merged.

AckslD commented on lua/neoclip/telescope.lua:
```diff
@@ -92,6 +105,26 @@ local function get_delete_handler(typ)
+                    regtype=entry.regtype,
+                }
+                storage.replace(typ, entry, new_entry)
+                resume_telescope_at_index(typ, telescope_opts, entry.index)
```
Not sure how difficult it would be but this might not open the picker at the edited entry if you did some yanks while editing the entry. I guess we could get the index from the sorted set but that's an O(n) operation.

Jeansidharta replied:
I really hadn't thought of that. I'll take a look in what can be done to solve this problem

Jeansidharta replied:
@AckslD I just changed the implementation to make it work in these situations. I ended up implementing the O(n) operation, but with a few caveats ([commit](https://github.com/AckslD/nvim-neoclip.lua/pull/93/commits/0761680e1fedc51fa1ca50ea5994f3cc64580bce) here)

In my implementation, I also provide an `index_hint`, which is the previous index of the entry. This index can be used as a starting point for the search. Therefore, the number of items searched until a match is the number of entries added to the list (worst case). This would not work if the user managed to remove an item from the list, in which case telescope would simply not pre-select the entry. But I couldn't think of a way for an item to be removed without nuking the database while still editing an entry.

However, I'm not really happy with this implementation. I only wrote this because it was the simplest way to fix the problem. I thought of some ways to improve this, but was not sure they'd be good implementations:
- **Better way to compare entries**: currently the only way to compare two entries is to use `sorted_set.hash` in both, and compare both hashes. This can be expensive, especially if the contents of the entries is large. A way to mitigate this is to add a proper id to the entries. The use of ids could even help if new features require changing the contents of the entry objects
- **Add an event system to `sorted_set`**: There could be a way for the `sorted_set` to let us know when a new entry has been added, and use this to update the index of the entry currently being edited. This seems a bit overengineering for a simple problem, but is a possible solution.

Anyway, these solutions seem a bit too complicated for this PR. I think the current implementation works and is not terrible. So I'd say we could leave it like that, and open another PR fixing this issue, if desired. What do you think?

AckslD replied:
Awesome @Jeansidharta! I think this seems fine also, especially since we'd only have to search through the new entries.

One easy optimization could maybe be to just add the hash to the entry so that it doesn't have to be re-computed every time. Then the hash function could use this if it's there. We would need to make sure it's updated/reset when we edit the entry though.

Anyway, I'm happy with this anyway :)

AckslD reviewed (commented):
Sorry for taking some time to reply. This is really cool!! :tada: 

Very well implemented. Just two comments and do you think we could add one test for doing the whole editing, ie opening the telescope picker, press `e` or `<c-e>` etc and see that the entry is the expected one? Basically a integration test.

AckslD commented:
> Hey, I noticed [this](https://github.com/AckslD/nvim-neoclip.lua/blob/main/lua/neoclip/telescope.lua#L57-L58) and [this](https://github.com/AckslD/nvim-neoclip.lua/blob/main/lua/neoclip/telescope.lua#L41-L42) todos could possibly be solved with [this](https://github.com/AckslD/nvim-neoclip.lua/pull/93/files#diff-1bdfb6249c123cbeb779faa6aa0f693fe2e2c19993cf1af4e4da224dd39ead8aR17-R29) new function that I just wrote. Instead of trying to actions these functions on the buffer "behind" telescope, we could close telescope, apply the actions, and reopen telescope. What do you think? I'd be willing to open another PR, after this one is merged.

I think that would be very nice, I think especially for the paste command it could be nice to paste more than one entries before closing the picker which is currently not possible.

Jeansidharta commented:
>

Jeansidharta commented:
> I think that would be very nice, I think especially for the paste command it could be nice to paste more than one entries before closing the picker which is currently not possible.

Cool. I'll wait for this to be merged and open another PR, just to separate both changes

Jeansidharta commented:
Hey @AckslD I just added some integration tests to this feature [here](https://github.com/AckslD/nvim-neoclip.lua/pull/93/commits/a82b40733ea90bd154858f992011fc43086578b0). Let me know if this is sufficient.

Also, I'll rename this PR to `feat: allow to edit and entry` for it to be more descriptive on what this is actually adding, and I'll rewrite it's description.

Jeansidharta commented:
Also, this should probably be marked as a breaking change, since it'd break the configs o users that were previously using the `<c-e>` mapping in telescope. But I have no idea how to mark this as such

AckslD commented:
Looks great @Jeansidharta! I think at least for `packer` and `lazy`, if you make a commit start with `feat!:` it will be flagged as a breaking change.

AckslD commented:
Could you maybe rebase/squash to make at least one of the commits have `feat!:`? Then I think this is ready to be merged :)

Jeansidharta commented:
> Could you maybe rebase/squash to make at least one of the commits have `feat!:`? Then I think this is ready to be merged :)

Awesome! I've squashed the commits, and added a proper commit message. I hope this is enough for users to know what's going on

AckslD commented:
Thanks! This is very nice! Maybe fun to announce this on reddit and/or [twin](https://this-week-in-neovim.org/). Maybe you'd like to do that since it's your addition :) Otherwise I can also do it if you want.

Jeansidharta commented:
Great Thank you a lot for your time and consideration. I know it must've taken you a while to review my code and answer my questions. I'm glad you liked the contribution.

I'll make a PR to TWiN announcing the changes, and will give a try at announcing it on Reddit. Though I'd like to run it by you before posting. I'll link it here for you to take a look :)
